server/src/main/java/org/elasticsearch/cluster/coordination/LagDetector.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    public void setTrackedNodes(final Iterable<DiscoveryNode> discoveryNodes) {
+        final Set<DiscoveryNode> discoveryNodeSet = new HashSet<>();
+        discoveryNodes.forEach(discoveryNodeSet::add);
+        appliedStateTrackersByNode.keySet().retainAll(discoveryNodeSet);
+        discoveryNodeSet.forEach(node -> appliedStateTrackersByNode.putIfAbsent(node, new NodeAppliedStateTracker(node)));
+    }
+
+    public void clearTrackedNodes() {
+        appliedStateTrackersByNode.clear();
+    }
+
+    public void setAppliedVersion(final DiscoveryNode discoveryNode, final long appliedVersion) {
+        final NodeAppliedStateTracker nodeAppliedStateTracker = appliedStateTrackersByNode.get(discoveryNode);
+        if (nodeAppliedStateTracker == null) {
+            logger.trace("node {} applied version {} but this node's version is not being tracked", discoveryNode, appliedVersion);
+        } else {
+            nodeAppliedStateTracker.increaseAppliedVersion(appliedVersion);
+        }
+    }
+
+    public void startLagDetector(final long version) {
+        appliedStateTrackersByNode.values().forEach(nodeAppliedStateTracker -> nodeAppliedStateTracker.startLagDetector(version));
+    }
+
+    @Override
+    public String toString() {
+        return "LagDetector{" +
+            "clusterStateApplicationTimeout=" + clusterStateApplicationTimeout +
+            ", appliedStateTrackersByNode=" + appliedStateTrackersByNode.values() +
+            '}';
+    }
+
+    // for assertions
+    Set<DiscoveryNode> getTrackedNodes() {
+        return Collections.unmodifiableSet(appliedStateTrackersByNode.keySet());
+    }
+
+    private class NodeAppliedStateTracker {
+        private final DiscoveryNode discoveryNode;
+        private final AtomicLong appliedVersion = new AtomicLong();
+
+        NodeAppliedStateTracker(final DiscoveryNode discoveryNode) {
+            this.discoveryNode = discoveryNode;
+        }
+
+        void increaseAppliedVersion(long appliedVersion) {
+            long maxAppliedVersion = this.appliedVersion.updateAndGet(v -> Math.max(v, appliedVersion));
+            logger.trace("{} applied version {}, max now {}", this, appliedVersion, maxAppliedVersion);
+        }
+
+        @Override
+        public String toString() {
+            return "NodeAppliedStateTracker{" +
+                "discoveryNode=" + discoveryNode +
+                ", appliedVersion=" + appliedVersion +
+                '}';
+        }
+
+        void startLagDetector(final long version) {
+            final long appliedVersionWhenStarted = appliedVersion.get();
+            if (version <= appliedVersionWhenStarted) {
+                logger.trace("lag detection for {} for version {} unnecessary, node has already applied version {}",
+                    discoveryNode, version, appliedVersionWhenStarted);
+                return;
+            }
+
+            threadPool.schedule(clusterStateApplicationTimeout, Names.GENERIC, new Runnable() {
```

Comment:
Even though we're running this only for nodes that took longer than Publish timeout, I'm still wondering if we can schedule a task at most once per publishing round, and not per node per publishing round.